cip/CIP2015-05-13-EXISTS.adoc

```diff
@@ -0,0 +1,154 @@
+= CIP2015-05-13 - EXISTENTIAL SUBQUERIES
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Authors:* Andrés Taylor <andres@neotechnology.com>
+
+
+[abstract]
+.Abstract
+--
+This CIP introduces EXISTS, a general function that can test for existence of properties, simple patterns and full subqueries.
+--
+
+toc::[]
+
+== Background
+
+In a way, Cypher already has existential subqueries - pattern predicates allow checking for the existence of subgraphs. This form is helpful, but not exhaustive in the forms of subqueries that users want to be able to express.
+
+== Proposal
+
+To make this feature more powerful, add a new EXISTS-function, that would exist in two forms:
+ . Property existence checking
+ . Subqueries existence checking
+
+=== Syntax
+
+----
+expression = …
+           | property exists
+           | subquery exists
+           | simple subquery exists
+            ;
+
+property exists = "EXISTS", "(", expression, ".", identifier ")"
+                | "EXISTS", "(", expression, "[", expression, "]", ")"
+                ;
+
+subquery exists = "EXISTS", "{", read only clause, { read only clause }, "}" ;
+
+simple subquery exists = "EXISTS", "{", simple match, "}" ;
+
+simple match = pattern, { ",", pattern }, [ "WHERE", predicate ] ;
+
+----
+Note: read only clause above includes WITH but not RETURN.
+
+=== Semantics
+
+All forms of +EXIST(...)+ accomplish the same task - check if a particular pattern exists in the database.
+They are expressions, and as such must be side-effect free. That is, the subqueries in +EXIST+ must not be updating queries.
+
+==== Property exists
+
+In the property checking form, either an identifier or a string expression is used to represent the property key name. If the node/relationship returned has a property with the given name, the +EXISTS(...)+ expression should return true, otherwise it should return false. If the node/relationship expression returns a null value, the +EXISTS(...)+ expression should return null as well.
+
+This form of +EXISTS+ replaces the already existing +has(...)+ function.
+
+==== Subquery exists
+Identifiers and parameters from the outer query are available in the subquery. Identifiers introduced in the subquery are not available on the outside.
+
```

Comment:
I guess we leave the term "identifiers" for now... even though we are looking to replace this with "variable"?